File: examples/typescript/basic_payment.ts
/**
 * KAMIYO PayAI Integration - Basic Payment Example (TypeScript)
 *
 * This example demonstrates how to:
 * 1. Request a paid endpoint
 * 2. Handle 402 Payment Required response
 * 3. Authorize payment via PayAI
 * 4. Retry with payment token
 * 5. Receive exploit intelligence data
 */

interface PaymentDetails {
  payment_required: boolean;
  endpoint: string;
  amount_usdc: number;
  payment_options: PaymentOption[];
}

interface PaymentOption {
  provider: string;
  type: string;
  priority: number;
  recommended: boolean;
  supported_chains?: string[];
  x402_standard?: X402Standard;
}

interface X402Standard {
  x402Version: number;
  error: string;
  accepts: PaymentAccept[];
}

interface PaymentAccept {
  scheme: string;
  network: string;
  maxAmountRequired: string;
  asset: string;
  payTo: string;
  resource: string;
  description: string;
}

interface ExploitData {
  exploits: Exploit[];
}

interface Exploit {
  tx_hash: string;
  chain: string;
  protocol?: string;
  amount_stolen_usd: number;
  attack_type?: string;
  timestamp: string;
}

// API Configuration
const API_BASE_URL = "https://api.kamiyo.ai";
const ENDPOINT = "/exploits";

/**
 * Request endpoint with automatic payment handling
 */
async function requestWithPayment(): Promise<ExploitData | null> {
  // Step 1: Request endpoint
  console.log(`📡 Requesting ${ENDPOINT}...`);
  let response = await fetch(`${API_BASE_URL}${ENDPOINT}`);

  // Step 2: Handle 402 Payment Required
  if (response.status === 402) {
    console.log("💰 Payment required!");
    const paymentDetails: PaymentDetails = await response.json();

    console.log(`   Amount: $${paymentDetails.amount_usdc} USDC`);
    console.log(`   Endpoint: ${paymentDetails.endpoint}`);
    console.log(`   Payment options: ${paymentDetails.payment_options.length}`);

    // Step 3: Get PayAI option
    const payaiOption = paymentDetails.payment_options.find(
      opt => opt.provider === "PayAI Network"
    );

    if (!payaiOption) {
      console.log("❌ PayAI option not available");
      return null;
    }

    console.log(`\n✅ Using PayAI Network`);
    console.log(`   Supported chains: ${payaiOption.supported_chains?.slice(0, 3).join(", ")}...`);

    // Step 4: Authorize payment
    const paymentToken = await authorizePayAIPayment(
      paymentDetails.endpoint,
      paymentDetails.amount_usdc,
      payaiOption
    );

    // Step 5: Retry with payment
    console.log(`\n🔄 Retrying with payment token...`);
    response = await fetch(`${API_BASE_URL}${ENDPOINT}`, {
      headers: {
        "X-PAYMENT": paymentToken
      }
    });

    if (response.status === 200) {
      const data: ExploitData = await response.json();
      console.log(`\n✅ Success! Received ${data.exploits.length} exploits`);
      return data;
    } else {
      console.log(`\n❌ Payment verification failed: ${response.status}`);
      return null;
    }
  } else if (response.status === 200) {
    // Already authenticated
    const data: ExploitData = await response.json();
    console.log(`✅ Success! Received ${data.exploits.length} exploits`);
    return data;
  } else {
    console.log(`❌ Error: ${response.status}`);
    return null;
  }
}

/**
 * Authorize payment via PayAI Network
 *
 * In production, replace this with actual PayAI SDK integration:
 * - Connect wallet (Phantom, MetaMask, etc.)
 * - Sign payment authorization
 * - Return base64-encoded payment token
 *
 * Example with PayAI SDK:
 * ```typescript
 * import { useX402 } from '@x402sdk/sdk';
 * const payment = await useX402(endpoint, amountUsdc, payaiOption.x402_standard);
 * return payment.token;
 * ```
 */
async function authorizePayAIPayment(
  endpoint: string,
  amountUsdc: number,
  payaiOption: PaymentOption
): Promise<string> {
  console.log(`\n💳 Authorizing payment...`);
  console.log(`   Endpoint: ${endpoint}`);
  console.log(`   Amount: $${amountUsdc} USDC`);

  // Mock payment authorization
  // In production, use PayAI SDK or wallet integration
  const mockPayment = {
    x402Version: 1,
    scheme: "exact",
    network: "base",
    payload: {
      payer: "0xMockAddress",
      amount: String(Math.floor(amountUsdc * 1e6)),  // USDC has 6 decimals
      resource: endpoint
    }
  };

  const paymentToken = btoa(JSON.stringify(mockPayment));
  console.log(`   ✅ Payment authorized`);

  return paymentToken;
}

/**
 * Main example function
 */
async function main() {
  console.log("=".repeat(70));
  console.log("KAMIYO PAYAI INTEGRATION - BASIC PAYMENT EXAMPLE");
  console.log("=".repeat(70) + "\n");

  const result = await requestWithPayment();

  if (result) {
    console.log("\n📊 Exploit Data:");
    result.exploits.slice(0, 3).forEach(exploit => {  // Show first 3
      console.log(`\n   🔴 ${exploit.protocol || "Unknown"}`);
      console.log(`      Amount: $${exploit.amount_stolen_usd.toLocaleString()}`);
      console.log(`      Chain: ${exploit.chain}`);
      console.log(`      Type: ${exploit.attack_type || "unknown"}`);
    });
  }

  console.log("\n" + "=".repeat(70));
}

// Run if executed directly
if (require.main === module) {
  main().catch(console.error);
}

export { requestWithPayment, authorizePayAIPayment };

File: examples/python/basic_payment.py
#!/usr/bin/env python3
"""
KAMIYO PayAI Integration - Basic Payment Example

This example demonstrates how to:
1. Request a paid endpoint
2. Handle 402 Payment Required response
3. Authorize payment via PayAI
4. Retry with payment token
5. Receive exploit intelligence data
"""

import httpx
import asyncio
from decimal import Decimal

# API Configuration
API_BASE_URL = "https://api.kamiyo.ai"
ENDPOINT = "/exploits"


async def request_with_payment():
    """Request paid endpoint with automatic payment handling"""

    async with httpx.AsyncClient() as client:
        # Step 1: Request endpoint
        print(f"📡 Requesting {ENDPOINT}...")
        response = await client.get(f"{API_BASE_URL}{ENDPOINT}")

        # Step 2: Handle 402 Payment Required
        if response.status_code == 402:
            print("💰 Payment required!")
            payment_details = response.json()

            print(f"   Amount: ${payment_details['amount_usdc']} USDC")
            print(f"   Endpoint: {payment_details['endpoint']}")
            print(f"   Payment options: {len(payment_details['payment_options'])}")

            # Step 3: Get PayAI option
            payai_option = next(
                opt for opt in payment_details['payment_options']
                if opt['provider'] == 'PayAI Network'
            )

            print(f"\n✅ Using PayAI Network")
            print(f"   Supported chains: {', '.join(payai_option['supported_chains'][:3])}...")

            # Step 4: Authorize payment (mock for example)
            # In production, use PayAI SDK or wallet integration
            payment_token = await authorize_payai_payment(
                payment_details['endpoint'],
                Decimal(str(payment_details['amount_usdc'])),
                payai_option
            )

            # Step 5: Retry with payment
            print(f"\n🔄 Retrying with payment token...")
            final_response = await client.get(
                f"{API_BASE_URL}{ENDPOINT}",
                headers={"X-PAYMENT": payment_token}
            )

            if final_response.status_code == 200:
                data = final_response.json()
                print(f"\n✅ Success! Received {len(data.get('exploits', []))} exploits")
                return data
            else:
                print(f"\n❌ Payment verification failed: {final_response.status_code}")
                return None

        elif response.status_code == 200:
            # Already authenticated
            data = response.json()
            print(f"✅ Success! Received {len(data.get('exploits', []))} exploits")
            return data

        else:
            print(f"❌ Error: {response.status_code}")
            return None


async def authorize_payai_payment(
    endpoint: str,
    amount_usdc: Decimal,
    payai_option: dict
) -> str:
    """
    Authorize payment via PayAI Network

    In production, replace this with actual PayAI SDK integration:
    - Connect wallet (Phantom, MetaMask, etc.)
    - Sign payment authorization
    - Return base64-encoded payment token

    Args:
        endpoint: API endpoint being accessed
        amount_usdc: Payment amount in USDC
        payai_option: PayAI payment option from 402 response

    Returns:
        Base64-encoded payment token
    """
    print(f"\n💳 Authorizing payment...")
    print(f"   Endpoint: {endpoint}")
    print(f"   Amount: ${amount_usdc} USDC")

    # Mock payment authorization
    # In production, use:
    # from payai_sdk import authorize_payment
    # token = await authorize_payment(endpoint, amount_usdc, payai_option)

    # For this example, return mock token
    import base64
    import json

    mock_payment = {
        "x402Version": 1,
        "scheme": "exact",
        "network": "base",
        "payload": {
            "payer": "0xMockAddress",
            "amount": str(int(amount_usdc * Decimal(10**6))),
            "resource": endpoint
        }
    }

    payment_token = base64.b64encode(
        json.dumps(mock_payment).encode('utf-8')
    ).decode('utf-8')

    print(f"   ✅ Payment authorized")
    return payment_token


async def main():
    """Main example function"""
    print("="*70)
    print("KAMIYO PAYAI INTEGRATION - BASIC PAYMENT EXAMPLE")
    print("="*70 + "\n")

    result = await request_with_payment()

    if result:
        print("\n📊 Exploit Data:")
        for exploit in result.get('exploits', [])[:3]:  # Show first 3
            print(f"\n   🔴 {exploit.get('protocol', 'Unknown')}")
            print(f"      Amount: ${exploit.get('amount_stolen_usd', 0):,.0f}")
            print(f"      Chain: {exploit.get('chain', 'unknown')}")
            print(f"      Type: {exploit.get('attack_type', 'unknown')}")

    print("\n" + "="*70)


if __name__ == "__main__":
    asyncio.run(main())
